src/StudyEntities/Candidate/IdentifierGenerator.php

```diff
@@ -0,0 +1,144 @@
+<?php declare(strict_types=1);
+
+namespace LORIS\StudyEntities\Candidate;
+
+abstract class IdentifierGenerator
+{
+    protected $generationMethod;
+    protected $alphabet = array();
+    protected $length;
+    protected $minValue;
+    protected $maxValue;
+    protected $prefix    = '';
+
+    /**
+    * Generates a new unused identifier to represent a Candidate in LORIS.
+    *
+    * @return string The new identifier
+    */
+    protected function createNewID(): string
+    {
+        $this->checkIDRangeFull();
+
+        $id = '';
+        switch ($this->generationMethod) {
+        case 'sequential':
+            $id = $this->padID($this->generateSequentialID());
+            break;
+        case 'random':
+            $id = $this->padID($this->generateRandomID());
+            // If the Random ID is already in use or is outside of the range 
+            // defined by minValue and maxValue, increment it by using the
+            // generateSequentialID function with the previous ID as a starting
+            // point.
+            if (in_array($id, $this->getExistingIDs(), true)
+                || strcmp(strval($id), strval($this->minValue)) < 0
+                || strcmp(strval($id), strval($this->maxValue)) > 0
+            ) {
+                $id = $this->generateSequentialID($id);
+            }
+            break;
+        }
+
+        // Return padded value with appropriate prefix
+        return $this->prefix . $id;
+    }
+
+    protected function generateSequentialID(string $id = ''): string
+    {
+        // If this is the first ID ever created, return the minimum value.
+        if (count($this->getExistingIDs()) < 1) {
+            return $this->minValue;
+        }
+        // Create the new ID by incrementing the value of the $id parameter OR
+        // by incrementing the highest existing ID.
+        $newID = !empty($id) ? $id : max($this->getExistingIDs());
+        // Increment newID until we find an unused value within the boundaries
+        // of $min and $max.
+        do {
+            // Make sure $newID is greater than or equal to minimum and less
+            // than or equal to maximum.
+            if (strcmp(strval($newID++), strval($this->minValue)) < 0) {
```

Comment:
if statements generally shouldn't have side-effects.

Moving $newID++ out of the if statement would avoid the need to figure out the order of operations when reviewing this code.